extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -0,0 +1,417 @@
+    }
+    return await OpenPGPKey.wrap(result.keys[0], {} as Key, text);
+  }
+
+  public static isPacketDecrypted = (pubkey: Key, keyid: string) => {
+    return OpenPGPKey.unwrap(pubkey).isPacketDecrypted({ bytes: keyid });
+  }
+
+  public static asPublicKey = async (pubkey: Key): Promise<Key> => {
+    if (pubkey.type !== 'openpgp') {
+      throw new Error('Unsupported key type: ' + pubkey.type);
+    }
+    if (pubkey.isPrivate) {
+      return await OpenPGPKey.wrap(OpenPGPKey.unwrap(pubkey).toPublic(), {} as Key);
+    }
+    return pubkey;
+  }
+
+  public static decryptKey = async (key: Key, passphrase: string, optionalKeyid?: string, optionalBehaviorFlag?: 'OK-IF-ALREADY-DECRYPTED'): Promise<boolean> => {
+    const prv = OpenPGPKey.unwrap(key);
+    if (!prv.isPrivate()) {
+      throw new Error("Nothing to decrypt in a public key");
+    }
+    const chosenPrvPackets = prv.getKeys(optionalKeyid ? { bytes: optionalKeyid } : undefined).map(k => k.keyPacket).filter(PgpKey.isPacketPrivate) as PrvPacket[];
+    if (!chosenPrvPackets.length) {
+      throw new Error(`No private key packets selected of ${prv.getKeys().map(k => k.keyPacket).filter(PgpKey.isPacketPrivate).length} prv packets available`);
+    }
+    for (const prvPacket of chosenPrvPackets) {
+      if (prvPacket.isDecrypted()) {
+        if (optionalBehaviorFlag === 'OK-IF-ALREADY-DECRYPTED') {
+          continue;
+        } else {
+          throw new Error("Decryption failed - key packet was already decrypted");
+        }
+      }
+      try {
+        await prvPacket.decrypt(passphrase); // throws on password mismatch
+      } catch (e) {
+        if (e instanceof Error && e.message.toLowerCase().includes('incorrect key passphrase')) {
+          return false;
+        }
+        throw e;
+      }
+    }
+    await OpenPGPKey.wrap(prv, key);
+    return true;
+  }
+
+  public static encryptKey = async (key: Key, passphrase: string) => {
+    const prv = OpenPGPKey.unwrap(key);
+    if (!passphrase || passphrase === 'undefined' || passphrase === 'null') {
+      throw new Error(`Encryption passphrase should not be empty:${typeof passphrase}:${passphrase}`);
+    }
+    const secretPackets = prv.getKeys().map(k => k.keyPacket).filter(PgpKey.isPacketPrivate);
+    const encryptedPacketCount = secretPackets.filter(p => !p.isDecrypted()).length;
+    if (!secretPackets.length) {
+      throw new Error(`No private key packets in key to encrypt. Is this a private key?`);
+    }
+    if (encryptedPacketCount) {
+      throw new Error(`Cannot encrypt a key that has ${encryptedPacketCount} of ${secretPackets.length} private packets still encrypted`);
+    }
+    await prv.encrypt(passphrase);
+    if (!prv.isFullyEncrypted()) {
+      throw new Error('Expected key to be fully encrypted after prv.encrypt');
+    }
+    await OpenPGPKey.wrap(prv, key);
+  }
+
+  public static decrypt = async (message: OpenPGP.message.Message, privateKeys: Key[], passwords?: string[]) => {
```

Comment:
Method `OpenPGPKey.decrypt` sounds like it would be decrypting a key, but it's actually decrypting a message

Reply:
Additionally, I remember this method being very large, but maybe I'm remembering another one - to look into

Reply:
Yes! This should be renamed as there already is `encryptKey` method. And yeah, there is one in pgp-msg but that got split into two that you see now (OpenPGP/SMIME).